server/src/test/java/org/opensearch/common/xcontent/JsonToStringXContentParserTests.java

```diff
@@ -0,0 +1,113 @@
+                parser,
+                fieldName
+            );
+            // Skip the START_OBJECT token:
+            jsonToStringXContentParser.nextToken();
+
+            XContentParser transformedParser = jsonToStringXContentParser.parseObject();
+            try (XContentBuilder jsonBuilder = XContentFactory.jsonBuilder()) {
+                jsonBuilder.copyCurrentStructure(transformedParser);
+                return jsonBuilder.toString();
+            }
+        }
+    }
+
+    public void testNestedObjects() throws IOException {
+        String jsonExample = "{" + "\"first\" : \"1\"," + "\"second\" : {" + "  \"inner\":  \"2.0\"" + "}," + "\"third\": \"three\"" + "}";
+
+        assertEquals(
+            "{"
+                + "\"flat\":[\"first\",\"second\",\"inner\",\"third\"],"
+                + "\"flat._value\":[\"1\",\"2.0\",\"three\"],"
+                + "\"flat._valueAndPath\":[\"flat.first=1\",\"flat.second.inner=2.0\",\"flat.third=three\"]"
+                + "}",
+            flattenJsonString("flat", jsonExample)
+        );
+    }
+
+    public void testChildHasDots() throws IOException {
+        // This should be exactly the same as testNestedObjects. We're just using the "flat" notation for the inner
+        // object.
+        String jsonExample = "{" + "\"first\" : \"1\"," + "\"second.inner\" : \"2.0\"," + "\"third\": \"three\"" + "}";
+
+        assertEquals(
+            "{"
+                + "\"flat\":[\"first\",\"second\",\"inner\",\"third\"],"
+                + "\"flat._value\":[\"1\",\"2.0\",\"three\"],"
+                + "\"flat._valueAndPath\":[\"flat.first=1\",\"flat.second.inner=2.0\",\"flat.third=three\"]"
+                + "}",
+            flattenJsonString("flat", jsonExample)
+        );
+    }
+
+    public void testNestChildObjectWithDots() throws IOException {
+        String jsonExample = "{"
+            + "\"first\" : \"1\","
+            + "\"second.inner\" : {"
+            + "  \"really_inner\" : \"2.0\""
+            + "},"
+            + "\"third\": \"three\""
+            + "}";
+
+        assertEquals(
+            "{"
+                + "\"flat\":[\"first\",\"second\",\"inner\",\"really_inner\",\"third\"],"
+                + "\"flat._value\":[\"1\",\"2.0\",\"three\"],"
+                + "\"flat._valueAndPath\":[\"flat.first=1\",\"flat.second.inner.really_inner=2.0\",\"flat.third=three\"]"
+                + "}",
+            flattenJsonString("flat", jsonExample)
+        );
+    }
+
+    public void testNestChildObjectWithDotsAndFieldWithDots() throws IOException {
+        String jsonExample = "{"
+            + "\"first\" : \"1\","
+            + "\"second.inner\" : {"
+            + "  \"totally.absolutely.inner\" : \"2.0\""
+            + "},"
+            + "\"third\": \"three\""
+            + "}";
+
+        assertEquals(
+            "{"
+                + "\"flat\":[\"first\",\"second\",\"inner\",\"totally\",\"absolutely\",\"inner\",\"third\"],"
```

Comment:
I'm not thrilled with how we're indexing the chain of keys, where we discard the path information.

If you look at this https://github.com/opensearch-project/OpenSearch/blob/69cc2a1b41a3e853675a9a2aa2ab47c9338bea37/rest-api-spec/src/main/resources/rest-api-spec/test/index/90_flat_object.yml#L521 test, we check for the existence of field `catalog.catalogId` and find it in the documents.

If we were to check for the existence of `catalog.author.surname`, we would not find it. If we tried to search for documents with `catalog.surname`, we *would* find them.

I think this behavior may be confusing. (It is the behavior we shipped, but I think it may have been a mistake.)

Reply:
@mingshl, @lukas-vlcek , as the original authors of the flat object feature, what do you think?

Reply:
I am thinking more about exist query, when I first wrote it, I was thinking of only using the most granular(one sub-field) of the dot path, 

for example, I want to check if this document has a subfield `author` or `surname`, but I don't remember the exact full path, 
it would be convenient to use exist query to look for a subfield name and see what is the exact dot path in the document. 
 
in your provided example,  `catalog` is a flat-object, so if I want to check `surname`, then I can search exist query with key `catalog.surname` and I don't have to memorize the exact full path. 

IMO, I think this is better, when there is deeply nested field, like `catalog.a.c.e.j.l.a.r.k.m`, and I want to check exist of subfield 'm', I can use exist query to search for `catalog.m` if exists.

If I know the full path, I can use `prefix` query instead.

Reply:
@reta @msfroh @lukas-vlcek this discussion of exist query in flat-object is not blocking this PR, it's more towards a feature discussion rather than a but, but would love to hear your opinions and we can make the change in a separate PR.

Reply:
> for example, I want to check if this document has a subfield author or surname, but I don't remember the exact full path,
it would be convenient to use exist query to look for a subfield name and see what is the exact dot path in the document.

this is good point, I think the exists query should be using full path like `catalog.a.c.e.j.l.a.r.k.m`, the reasoning is that  someone expects the JSON to follow a particular structure (not just have `m` somewhere but exactly under `catalog.a.c.e.j.l.a.r.k`).

Reply:
Okay -- I'll address this in a follow-up issue. (I'll probably edit this comment with the link to that issue once I've written it.)